### PR TITLE
Draft: `/messages` investigation scratch pad1

### PR DESCRIPTION
**Pull request by MadLittleMods**
Part of https://github.com/matrix-org/synapse/issues/13356

Combine:

 - OTEL migration https://github.com/matrix-org/synapse/pull/13400
 - with `/messages` `@trace` decorations, https://github.com/matrix-org/synapse/pull/13368

So that I can run against Complement federation tests and see if there is more to add `@trace` to in the federation stack of things when `/messages` happens.

#### We load a lot of state (from 2. in https://github.com/matrix-org/synapse/issues/13356)

In `#matrixhq` there are 40k current members and I assume `get_current_state` is the root cause why we `Loaded 79277 events` (seems like that took 17s too). We only call `get_current_state` in order to get a list of likely domains to backfill from.

We could optimize this by:

 1. Cache `get_domains_from_state` so we don't have to `get_current_state` as much
 1. Use a smarter heuristic to get the first couple domains to try to fetch from and do the `get_domains_from_state` in the background so it's ready by the time we fail with the first couple of domains.

#### Skip backfill

Skip backfill or kick it off in the background if it's not our first time and we have enough events.

We don't want to get stuck on the same unfetchable event over and over.

#### Why is `/state_ids` slow to respond?

We can't control every bad network effect but maybe Synapse is slow to assemble a `/state_ids` reponse 🤔 Need to investigate

`FederationStateIdsServlet`

 - Sometimes it's slow because the ratelimiter kicks in (traced in https://github.com/matrix-org/synapse/pull/13499)

#### We should only care about `auth_event_ids`

We should only care about getting the `event_id` and `auth_event_ids` in `_get_state_ids_after_missing_prev_event(...)`

We shouldn't factor  `state_event_ids` into whether

### Dev notes

```
TEST_ONLY_IGNORE_POETRY_LOCKFILE=1 TEST_ONLY_SKIP_DEP_HASH_VERIFICATION=1 COMPLEMENT_ALWAYS_PRINT_SERVER_LOGS=1 COMPLEMENT_DIR=../complement ./scripts-dev/complement.sh -run TestMessagesOverFederation -p 1
```

```
TEST_ONLY_IGNORE_POETRY_LOCKFILE=1 TEST_ONLY_SKIP_DEP_HASH_VERIFICATION=1 COMPLEMENT_ALWAYS_PRINT_SERVER_LOGS=1 COMPLEMENT_DIR=../complement ./scripts-dev/complement.sh -run TestImportHistoricalMessages/parallel/Federation/Backfill_still_works_after_many_batches_are_imported -p 1
```

```
poetry run synapse_homeserver --config-path homeserver.yaml
```

#### Jaeger max duration spans

`213503982d 8h`, see https://github.com/matrix-org/synapse/pull/13440#discussion_r940725431

### Pull Request Checklist

* [ ] Pull request is based on the develop branch
* [ ] Pull request includes a [changelog file](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
  - Feel free to credit yourself, by adding a sentence "Contributed by @github_username." or "Contributed by [Your Name]." to the end of the entry.
* [ ] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
* [ ] [Code style](https://matrix-org.github.io/synapse/latest/code_style.html) is correct
  (run the [linters](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#run-the-linters))

### CODE REVIEW

File: docker/conf/homeserver.yaml [hunk 1]
```diff
@@ -184,3 +184,21 @@ trusted_key_servers:
...
+    agent_host_name: host.docker.internal
+    agent_port: 6831
+    # Split UDP packets (UDP_PACKET_MAX_LENGTH is set to 65k in OpenTelemetry)
+    udp_split_oversized_batches: true
+    # If you define a collector, it will communicate directly to the collector,
+    # bypassing the agent
+    #
+    # It does not seem like the agent can keep up with the massive UDP load
+    # (1065 spans in one trace) so lets just use the HTTP collector endpoint
+    # instead which seems to work.
```
Comment (MadLittleMods) [inline docker/conf/homeserver.yaml:203]: I wonder why this is the case? I was seeing this same behavior with the Jaeger `opentracing` stuff. Is the UDP connection being over saturated? Can the Jaeger agent not keep up? We see some spans come over but never the main servlet overarching one that is probably the last to be exported.

But using the HTTP Jaeger collector endpoint seems to work fine for getting the whole trace.

File: docker/conf/homeserver.yaml [hunk 2]
```diff
@@ -184,3 +184,21 @@ trusted_key_servers:
 
 password_config:
```
Comment (MadLittleMods) [inline docker/conf/homeserver.yaml:185]: Occasionally, I see spans that have a super massive duration and shows up with `213503982d 8h` in the Jaeger UI. I only see one other reference to this magic value in this issue https://github.com/jaegertracing/jaeger-ui/issues/397#issuecomment-937704585 which they describe as the "max duration".



 
I can't figure out the math that makes this situation happen. Whether it be forgotten `end_time`, `start_time`, max int, infinity, etc. The exporter shouldn't get any non-ended span so it seems like the `end_time` is being set to some astronomical value somehow 🤔

The raw data for those spans are:

`db.get_partial_state_events`
```jsonc
{
    "traceID": "563478e5d75db03e9fc028822fde0649",
    "spanID": "5b63b0172763c5cf",
    "flags": 1,
    "operationName": "db.get_partial_state_events",
    "references": [
        {
            "refType": "CHILD_OF",
            "traceID": "563478e5d75db03e9fc028822fde0649",
            "spanID": "10c9f2a2c50e286f"
        }
    ],
    "startTime": 1659994809436162,
    "duration": 18446744073709517944,
    "tags": [
        // ...
    ],
    "logs": [],
    "processID": "p1",
    "warnings": null
}
```

`_process_pulled_event`
```
{
    "traceID": "563478e5d75db03e9fc028822fde0649",
    "spanID": "6e9ee2608b03b542",
    "flags": 1,
    "operationName": "_process_pulled_event",
    "references": [
        {
            "refType": "CHILD_OF",
            "traceID": "563478e5d75db03e9fc028822fde0649",
            "spanID": "af2c9c8d08284134"
        }
    ],
    "startTime": 1659994809432356,
    "duration": 18446744073709532019,
    "tags": [
        // ...
    ],
    "logs": [],
    "processID": "p1",
    "warnings": null
},
```

Trying to work out the math to get there

```
213503982d 8h (displayed in Jaeger UI)
5124095576h
18446744073600s
18446744073600000ms
18446744073600000000us
```

```
_process_pulled_event
1659994809432356        (start_time)
18446744073709532019    (duration)
18448404068518965000    (end_time, derived)

db.get_partial_state_events
1659994809436162        (start_time)
18446744073709517944    (duration)
18448404068518957000    (end_time, derived)
```

[`start_time` and `end_time` are calculated by `time.time_ns()`](https://github.com/open-telemetry/opentelemetry-python/blob/1cae35f5ad3c5beabc067e57a3940d3dfeaac8e3/opentelemetry-sdk/src/opentelemetry/sdk/trace/__init__.py#L867-L880) ([`_time_ns`](https://github.com/open-telemetry/opentelemetry-python/blob/1cae35f5ad3c5beabc067e57a3940d3dfeaac8e3/opentelemetry-api/src/opentelemetry/util/_time.py#L34))

And the Jaeger exporter calculates `startTime` and `duration` with this small [`_nsec_to_usec_round`](https://github.com/open-telemetry/opentelemetry-python/blob/1cae35f5ad3c5beabc067e57a3940d3dfeaac8e3/exporter/opentelemetry-exporter-jaeger-thrift/src/opentelemetry/exporter/jaeger/thrift/translate/__init__.py#L164-L165) function.

**Comment by clokep** [comment]
@MadLittleMods Is this useful or have you gleaned everything you can from it?

File: docker/conf/homeserver.yaml (hunk 2 above)
Comment (MadLittleMods) [reply docker/conf/homeserver.yaml:185]: Definitely close to the Python max int size (unsigned word, 64 bits): `sys.maxsize * 2 + 1` -> `18446744073709551615` (via https://stackoverflow.com/questions/7604966/maximum-and-minimum-values-for-ints)

Comment (MadLittleMods) [reply docker/conf/homeserver.yaml:185]: Conversation moved to https://github.com/matrix-org/synapse/pull/13400#discussion_r1199505260

**Pull request closed without merging**